File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def dfs(start, sum, path):
            if sum == target:
                result.append(path[:])
                return

            for i in range(start, len(candidates)):
                path.append(candidates[i])
                sum += candidates[i]

                if sum <= target:
                    dfs(i, sum, path)

                sum -= candidates[i]
                path.pop()

        result = []
        dfs(0, 0, [])

        return result

File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, target) {
    const dfs = function (start, sum, path) {
        if (sum === target) {
            result.push([...path]);
            return;
        }

        for (let i = start; i < candidates.length; i++) {
            path.push(candidates[i]);
            sum += candidates[i];

            if(sum <= target) {
                dfs(i, sum, path);
            }

            sum -= candidates[i];
            path.pop();
        }
    }

    const result = [];
    dfs(0, 0, []);

    return result;
};